flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/SnapshotTriggerTimestampStore.java

```diff
@@ -0,0 +1,133 @@
+import static org.apache.flink.kubernetes.operator.api.status.SnapshotTriggerType.PERIODIC;
+import static org.apache.flink.kubernetes.operator.reconciler.SnapshotType.SAVEPOINT;
+
+/** Class used to store latest timestamps of periodic checkpoint/savepoint. */
+@RequiredArgsConstructor
+public class SnapshotTriggerTimestampStore {
+    private final ConcurrentHashMap<String, Instant> checkpointsLastTriggeredCache =
+            new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, Instant> savepointsLastTriggeredCache =
+            new ConcurrentHashMap<>();
+
+    /**
+     * Returns the time a periodic snapshot was last triggered for this resource. This is stored in
+     * memory, on operator start it will use the latest completed FlinkStateSnapshot CR creation
+     * timestamp. If none found, the return value will be the max of the resource creation timestamp
+     * and the latest triggered legacy snapshot.
+     *
+     * @param resource Flink resource
+     * @param snapshotsSupplier supplies related snapshot resources
+     * @return instant of last trigger
+     */
+    public Instant getLastPeriodicTriggerInstant(
+            AbstractFlinkResource<?, ?> resource,
+            SnapshotType snapshotType,
+            Supplier<Set<FlinkStateSnapshot>> snapshotsSupplier) {
+        var cache = getCacheForSnapshotType(snapshotType);
+        if (cache.containsKey(resource.getMetadata().getUid())) {
+            return cache.get(resource.getMetadata().getUid());
+        }
+
+        var snapshots = Optional.ofNullable(snapshotsSupplier.get()).orElse(Set.of());
+        return snapshots.stream()
+                .filter(s -> s.getStatus() != null && COMPLETED.equals(s.getStatus().getState()))
+                .filter(s -> (snapshotType == SAVEPOINT) == s.getSpec().isSavepoint())
+                .filter(
+                        s ->
+                                PERIODIC.name()
+                                        .equals(
+                                                s.getMetadata()
+                                                        .getLabels()
+                                                        .get(CrdConstants.LABEL_SNAPSHOT_TYPE)))
+                .map(s -> DateTimeUtils.parseKubernetes(s.getMetadata().getCreationTimestamp()))
+                .max(Comparator.naturalOrder())
+                .orElseGet(
+                        () -> {
+                            var legacyTs = getLegacyTimestamp(resource, snapshotType);
+                            var creationTs =
+                                    Instant.parse(resource.getMetadata().getCreationTimestamp());
+
+                            if (legacyTs.compareTo(creationTs) > 0) {
+                                return legacyTs;
+                            } else {
+                                return creationTs;
+                            }
+                        });
```

Comment:
Should we put into the cache before returning?

Reply:
I like this idea, with this change we won't unnecessarily call the supplier if we don't trigger a periodic snapshot in AbstractJobReconciler.